File: libraries/typescript/src/embedded_counter_cache.test.ts
import { EmbeddedCounterCache } from './embedded_counter_cache';
import { AeronCacheClient } from './index';

jest.mock('./index');

describe('EmbeddedCounterCache', () => {
    let mockClient: jest.Mocked<AeronCacheClient>;
    let cache: EmbeddedCounterCache;

    beforeEach(() => {
        mockClient = new AeronCacheClient('http://test', 'ws://test') as jest.Mocked<AeronCacheClient>;
        cache = new EmbeddedCounterCache(mockClient, 'test-counter');
    });

    it('delegates put to client', async () => {
        mockClient.putCounter.mockResolvedValue({ operationStatus: 'SUCCESS' } as any);
        await cache.put('hits', 42);
        expect(mockClient.putCounter).toHaveBeenCalledWith('test-counter', 'hits', 42);
    });

    it('delegates increment/decrement/set to client', async () => {
        mockClient.incrementCounter.mockResolvedValue({ value: 43 } as any);
        mockClient.decrementCounter.mockResolvedValue({ value: 41 } as any);
        mockClient.setCounter.mockResolvedValue({ value: 100 } as any);

        await cache.increment('hits', 1);
        await cache.decrement('hits', 1);
        await cache.set('hits', 100);

        expect(mockClient.incrementCounter).toHaveBeenCalledWith('test-counter', 'hits', 1);
        expect(mockClient.decrementCounter).toHaveBeenCalledWith('test-counter', 'hits', 1);
        expect(mockClient.setCounter).toHaveBeenCalledWith('test-counter', 'hits', 100);
    });

    it('delegates clear to client', async () => {
        mockClient.deleteCounterCache.mockResolvedValue({ operationStatus: 'SUCCESS' } as any);
        await cache.clear();
        expect(mockClient.deleteCounterCache).toHaveBeenCalledWith('test-counter');
    });

    it('updates local cache via subscription and returns local value', () => {
        let subscriptionCallback: any;
        const mockWs = { close: jest.fn() };

        mockClient.subscribeCounter.mockImplementation((cacheId, onMessage) => {
            subscriptionCallback = onMessage;
            return mockWs;
        });

        cache.subscribe(jest.fn());

        subscriptionCallback({ eventType: 'ADD_ITEM', itemKey: 'hits', itemValue: 42 });
        expect(cache.getLocal('hits')).toBe(42);

        // Zero must be honoured (not treated as missing)
        subscriptionCallback({ eventType: 'ADD_ITEM', itemKey: 'hits', itemValue: 0 });
        expect(cache.getLocal('hits')).toBe(0);

        subscriptionCallback({ eventType: 'REMOVE_ITEM', itemKey: 'hits' });
        expect(cache.getLocal('hits')).toBeUndefined();
    });
});

File: libraries/java/src/test/java/com/aeron/cache/client/EmbeddedCounterCacheTest.java
package com.aeron.cache.client;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;
import java.net.http.WebSocket;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.*;

public class EmbeddedCounterCacheTest {

    private AeronCacheClient mockClient;
    private EmbeddedCounterCache cache;

    @BeforeEach
    void setUp() {
        mockClient = mock(AeronCacheClient.class);
        cache = new EmbeddedCounterCache(mockClient, "test-counter");
    }

    @Test
    public void testPutDelegatesToClient() throws Exception {
        cache.put("hits", 42);
        verify(mockClient, times(1)).putCounter("test-counter", "hits", 42L);
    }

    @Test
    public void testIncrementDelegatesToClient() throws Exception {
        cache.increment("hits", 1);
        verify(mockClient, times(1)).incrementCounter("test-counter", "hits", 1L);
    }

    @Test
    public void testDecrementDelegatesToClient() throws Exception {
        cache.decrement("hits", 1);
        verify(mockClient, times(1)).decrementCounter("test-counter", "hits", 1L);
    }

    @Test
    public void testSetDelegatesToClient() throws Exception {
        cache.set("hits", 100);
        verify(mockClient, times(1)).setCounter("test-counter", "hits", 100L);
    }

    @Test
    public void testClearDelegatesToClient() throws Exception {
        cache.clear();
        verify(mockClient, times(1)).deleteCounterCache("test-counter");
    }

    @Test
    public void testUpdatesLocalCacheViaSubscription() {
        ReconnectingWebSocket mockWebSocket = mock(ReconnectingWebSocket.class);
        ArgumentCaptor<CounterCacheSubscriber> captor = ArgumentCaptor.forClass(CounterCacheSubscriber.class);
        when(mockClient.subscribeCounter(eq("test-counter"), captor.capture())).thenReturn(mockWebSocket);

        cache.subscribe(new CounterCacheSubscriber() {});

        CounterCacheSubscriber registeredSubscriber = captor.getValue();
        WebSocket baseMockWs = mock(WebSocket.class);

        // Simulate ADD_ITEM
        registeredSubscriber.onText(baseMockWs, "{\"eventType\":\"ADD_ITEM\",\"itemKey\":\"hits\",\"itemValue\":42}", true);
        assertEquals(Long.valueOf(42L), cache.getLocal("hits"));

        // Zero is a valid counter value and must be stored, not treated as missing.
        registeredSubscriber.onText(baseMockWs, "{\"eventType\":\"ADD_ITEM\",\"itemKey\":\"hits\",\"itemValue\":0}", true);
        assertEquals(Long.valueOf(0L), cache.getLocal("hits"));

        // Simulate REMOVE_ITEM
        registeredSubscriber.onText(baseMockWs, "{\"eventType\":\"REMOVE_ITEM\",\"itemKey\":\"hits\"}", true);
        assertNull(cache.getLocal("hits"));
    }
}
